// ion-alpha-scroll/dist/index.js
"use strict";
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var common_1 = require("@angular/common");
var alpha_scroll_component_1 = require("./alpha-scroll.component");
var alpha_scroll_service_1 = require("./alpha-scroll.service");
var order_by_1 = require("./order-by");
var ionic_angular_1 = require("ionic-angular");
var alpha_scroll_component_2 = require("./alpha-scroll.component");
exports.AlphaScrollComponent = alpha_scroll_component_2.AlphaScrollComponent;
var alpha_scroll_service_2 = require("./alpha-scroll.service");
exports.AlphaScrollService = alpha_scroll_service_2.AlphaScrollService;
var AlphaScrollModule = /** @class */ (function () {
    function AlphaScrollModule() {
    }
    /**
     * @return {?}
     */
    AlphaScrollModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: AlphaScrollModule,
            providers: [order_by_1.OrderBy, alpha_scroll_service_1.AlphaScrollService]
        };
    };
    AlphaScrollModule.decorators = [
        { type: core_1.NgModule, args: [{
                    imports: [
                        common_1.CommonModule,
                        ionic_angular_1.IonicModule
                    ],
                    declarations: [
                        alpha_scroll_component_1.AlphaScrollComponent,
                        order_by_1.OrderBy
                    ],
                    exports: [
                        alpha_scroll_component_1.AlphaScrollComponent,
                        order_by_1.OrderBy
                    ]
                },] },
    ];
    return AlphaScrollModule;
}());
exports.AlphaScrollModule = AlphaScrollModule;
function AlphaScrollModule_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    AlphaScrollModule.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    AlphaScrollModule.ctorParameters;
}
